File: apps/Model Plane/go/services/cost-core/internal/pricing/pricing.go
// Package pricing turns a reported model name + token counts into a USD cost.
//
// cost-core records what it is told, but most usage events arrive carrying only
// token counts (the gateway publishes tokens; cost was historically left at 0,
// which made the dollar ledger — and therefore the budget posture — always
// read $0). The resolver closes that gap: it loads a price catalogue
// (model_pricing table, seeded by migration 0002) and computes cost when a
// usage event has none.
//
// Matching is forgiving so a concrete deployment name still prices: an exact
// match wins, else the longest model key that is a prefix of the reported name
// (so `claude-sonnet-4-6` → `claude-sonnet`, `gpt-4o-mini-2024-07-18` →
// `gpt-4o-mini`), else the mandatory `default` row. The gateway mirrors this
// logic for the SSE display value off the same catalogue (GET /api/v1/pricing),
// so the streamed cost and the ledgered cost agree.
package pricing

import (
	"context"
	"sort"
	"strings"

	"github.com/jackc/pgx/v5/pgxpool"
)

// DefaultModelKey is the mandatory fallback key. It is always present so an
// unrecognised model is priced (at a mid tier) rather than counted as free.
const DefaultModelKey = "default"

// CacheReadDiscount is the fraction of the base input-token rate a
// prompt-cache read is billed at -- the documented 10x discount Anthropic
// (and other providers with prompt caching) apply to a cache hit. Mirrored
// exactly in the gateway's Rust pricing cache
// (services/model-gateway/src/pricing.rs) so the streamed display cost and
// this ledgered cost agree.
const CacheReadDiscount = 0.1

// Rate is the price for one model key, in USD per 1,000,000 tokens.
type Rate struct {
	Model            string  `json:"model"`
	InputPerMillion  float64 `json:"input_per_million"`
	OutputPerMillion float64 `json:"output_per_million"`
	Currency         string  `json:"currency"`
}

// Resolver maps a model name to a Rate and computes USD cost. It is read-only
// after construction (safe for concurrent use).
type Resolver struct {
	rates map[string]Rate // keyed by normalized model key
	def   Rate            // the DefaultModelKey rate
}

// defaultSeed mirrors migrations/0002_model_pricing.up.sql so the in-memory /
// connect-failure path prices identically to the durable catalogue.
var defaultSeed = []Rate{
	{Model: DefaultModelKey, InputPerMillion: 3.00, OutputPerMillion: 15.00},
	{Model: "gpt-4o-mini", InputPerMillion: 0.15, OutputPerMillion: 0.60},
	{Model: "gpt-4o", InputPerMillion: 2.50, OutputPerMillion: 10.00},
	{Model: "gpt-4.1-mini", InputPerMillion: 0.40, OutputPerMillion: 1.60},
	{Model: "gpt-4.1", InputPerMillion: 2.00, OutputPerMillion: 8.00},
	{Model: "o4-mini", InputPerMillion: 1.10, OutputPerMillion: 4.40},
	{Model: "o3", InputPerMillion: 2.00, OutputPerMillion: 8.00},
	{Model: "model-router", InputPerMillion: 2.00, OutputPerMillion: 8.00},
	{Model: "claude-haiku", InputPerMillion: 0.80, OutputPerMillion: 4.00},
	{Model: "claude-3-5-haiku", InputPerMillion: 0.80, OutputPerMillion: 4.00},
	{Model: "claude-sonnet", InputPerMillion: 3.00, OutputPerMillion: 15.00},
	{Model: "claude-3-5-sonnet", InputPerMillion: 3.00, OutputPerMillion: 15.00},
	{Model: "claude-opus", InputPerMillion: 15.00, OutputPerMillion: 75.00},
	{Model: "claude-3-opus", InputPerMillion: 15.00, OutputPerMillion: 75.00},
}

// Default builds a resolver from the built-in seed. Used when no database is
// configured and as a fallback when loading the catalogue fails.
func Default() *Resolver {
	return fromRates(defaultSeed)
}

// LoadFromPool reads the model_pricing catalogue from Postgres. If the table is
// empty or missing a default row, the built-in seed backfills it so a default
// rate is always available.
func LoadFromPool(ctx context.Context, pool *pgxpool.Pool) (*Resolver, error) {
	rows, err := pool.Query(ctx, `
		SELECT model, input_per_million::double precision,
		       output_per_million::double precision, currency
		FROM model_pricing`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var loaded []Rate
	for rows.Next() {
		var r Rate
		if err := rows.Scan(&r.Model, &r.InputPerMillion, &r.OutputPerMillion, &r.Currency); err != nil {
			return nil, err
		}
		loaded = append(loaded, r)
	}
	if err := rows.Err(); err != nil {
		return nil, err
	}
	// Backfill from the seed for any key the DB lacks (notably `default`).
	seen := make(map[string]struct{}, len(loaded))
	for _, r := range loaded {
		seen[normalize(r.Model)] = struct{}{}
	}
	for _, r := range defaultSeed {
		if _, ok := seen[normalize(r.Model)]; !ok {
			loaded = append(loaded, r)
		}
	}
	return fromRates(loaded), nil
}

func fromRates(rates []Rate) *Resolver {
	r := &Resolver{rates: make(map[string]Rate, len(rates))}
	for _, rate := range rates {
		if rate.Currency == "" {
			rate.Currency = "USD"
		}
		key := normalize(rate.Model)
		rate.Model = key
		r.rates[key] = rate
		if key == DefaultModelKey {
			r.def = rate
		}
	}
	if r.def.Model == "" {
		r.def = Rate{Model: DefaultModelKey, InputPerMillion: 3.00, OutputPerMillion: 15.00, Currency: "USD"}
		r.rates[DefaultModelKey] = r.def
	}
	return r
}

// Cost returns the USD cost of an inference given its model and token counts.
// Negative token counts are clamped to zero.
func (r *Resolver) Cost(model string, inputTokens, outputTokens int64) float64 {
	return r.CostWithCache(model, inputTokens, outputTokens, 0, 0)
}

// CostWithCache is [Resolver.Cost] plus prompt-cache telemetry: a cache-token
// telemetry prerequisite for the native-compaction migration, so cache-hit
// rate and its cost savings are visible on the durable ledger.
//
// cacheReadInputTokens and cacheCreationInputTokens are cache legs already
// folded into inputTokens by the serving adapter (Anthropic's
// `total_input_tokens` fold-in, mirrored on the Rust side) -- passing 0 for
// both reproduces Cost's result exactly, so every existing caller (and every
// non-Anthropic provider today) is unaffected byte-for-byte. A cache-read leg
// is billed at CacheReadDiscount (0.1x) of the base input rate instead of the
// full rate: inputTokens is charged in full first, then the cache-read leg's
// over-charge is credited back. A cache-creation leg stays at the full input
// rate (Anthropic writes it fresh, no discount) -- it is already counted
// correctly inside inputTokens and needs no adjustment; the parameter exists
// so a future differential rate has one call site to change.
func (r *Resolver) CostWithCache(model string, inputTokens, outputTokens, cacheReadInputTokens, _cacheCreationInputTokens int64) float64 {
	rate := r.lookup(model)
	in := float64(max64(inputTokens, 0))
	out := float64(max64(outputTokens, 0))
	cacheRead := float64(max64(cacheReadInputTokens, 0))
	return in/1_000_000.0*rate.InputPerMillion +
		out/1_000_000.0*rate.OutputPerMillion -
		cacheRead/1_000_000.0*rate.InputPerMillion*(1.0-CacheReadDiscount)
}

// lookup resolves a model name to a Rate: exact key, else longest prefix key,
// else the default.
func (r *Resolver) lookup(model string) Rate {
	key := normalize(model)
	if key == "" {
		return r.def
	}
	if rate, ok := r.rates[key]; ok {
		return rate
	}
	var best Rate
	bestLen := -1
	for candidate, rate := range r.rates {
		if candidate == DefaultModelKey {
			continue
		}
		if strings.HasPrefix(key, candidate) && len(candidate) > bestLen {
			best = rate
			bestLen = len(candidate)
		}
	}
	if bestLen >= 0 {
		return best
	}
	return r.def
}

// Rates returns the full catalogue, sorted by model key (default first), for
// the GET /api/v1/pricing endpoint.
func (r *Resolver) Rates() []Rate {
	out := make([]Rate, 0, len(r.rates))
	for _, rate := range r.rates {
		out = append(out, rate)
	}
	sort.Slice(out, func(i, j int) bool {
		if out[i].Model == DefaultModelKey {
			return true
		}
		if out[j].Model == DefaultModelKey {
			return false
		}
		return out[i].Model < out[j].Model
	})
	return out
}

func normalize(model string) string {
	return strings.ToLower(strings.TrimSpace(model))
}

func max64(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

File: apps/Model Plane/rust/services/model-gateway/src/pricing.rs
//! Model pricing cache — turns reported token counts into a USD cost for the
//! streamed `Usage` SSE event, using cost-core's price catalogue as the single
//! source of truth.
//!
//! cost-core prices the durable ledger authoritatively (so the dollar ledger
//! and budget posture are real); this cache lets the SSE show the *same* figure
//! without a per-request hop to the database. It fetches `GET /api/v1/pricing`
//! once and refreshes lazily once the TTL lapses. If cost-core is unreachable
//! and the cache is cold, [`PricingCache::cost_usd`] returns `None` — the
//! gateway emits a null cost rather than a fabricated one.
//!
//! Matching mirrors cost-core's resolver: exact key, else the longest model key
//! that is a prefix of the reported name, else the mandatory `default` row.

use std::sync::Arc;
use std::time::{Duration, Instant};

use serde::Deserialize;
use tokio::sync::RwLock;

const DEFAULT_MODEL_KEY: &str = "default";
const REFRESH_TTL: Duration = Duration::from_secs(600);
const FETCH_TIMEOUT: Duration = Duration::from_secs(2);

/// A prompt-cache read is billed at this fraction of the base input-token
/// rate -- the documented 10x discount Anthropic (and the other providers
/// that support prompt caching) apply to a cache hit. Mirrored exactly in
/// cost-core's Go resolver (`services/cost-core/internal/pricing`) so the
/// streamed display cost and the ledgered cost agree.
const CACHE_READ_DISCOUNT: f64 = 0.1;

#[derive(Clone, Debug, Deserialize)]
struct Rate {
    model: String,
    #[serde(default)]
    input_per_million: f64,
    #[serde(default)]
    output_per_million: f64,
}

#[derive(Deserialize)]
struct PricingResponse {
    #[serde(default)]
    rates: Vec<Rate>,
}

#[derive(Default)]
struct CacheInner {
    /// Catalogue keyed by a normalized (lowercased, trimmed) model key.
    rates: Vec<Rate>,
    fetched_at: Option<Instant>,
}

/// Shared, cheaply-cloneable pricing cache (Arc-backed).
#[derive(Clone)]
pub struct PricingCache {
    /// cost-core base URL (e.g. `http://model-plane-cost-core-1:8089`). `None`
    /// disables pricing — `cost_usd` then always returns `None`.
    base_url: Option<String>,
    http: reqwest::Client,
    inner: Arc<RwLock<CacheInner>>,
}

impl PricingCache {
    /// Build a cache against cost-core's base URL. An empty/blank URL disables
    /// pricing (the gateway then emits a null cost — never a fake one).
    #[must_use]
    pub fn new(base_url: Option<String>, http: reqwest::Client) -> Self {
        let base_url = base_url
            .map(|u| u.trim().trim_end_matches('/').to_owned())
            .filter(|u| !u.is_empty());
        Self {
            base_url,
            http,
            inner: Arc::new(RwLock::new(CacheInner::default())),
        }
    }

    /// Compute the USD cost for an inference. Returns `None` when no catalogue
    /// is available (pricing disabled, or cost-core unreachable with a cold
    /// cache) so the caller can emit a null `cost_usd` instead of faking one.
    ///
    /// `cache_read_input_tokens` and `cache_creation_input_tokens` are the
    /// same cache legs already folded into `input_tokens` (see
    /// `inference-core`'s `total_input_tokens`): pass 0 for a request that
    /// reported no cache usage, which reproduces the pre-cache-pricing
    /// result exactly. A cache-read leg is billed at `CACHE_READ_DISCOUNT`
    /// (0.1x) of the base input rate rather than the full rate. A
    /// cache-creation leg is billed at the full input rate like ordinary
    /// input (Anthropic writes it fresh, so no discount applies) — it is
    /// already counted correctly inside `input_tokens` and needs no separate
    /// adjustment here; the parameter exists so a future differential rate
    /// has one call site to change instead of a signature to add.
    pub async fn cost_usd(
        &self,
        model: &str,
        input_tokens: i64,
        output_tokens: i64,
        cache_read_input_tokens: i64,
        _cache_creation_input_tokens: i64,
    ) -> Option<f64> {
        self.ensure_fresh().await;
        let inner = self.inner.read().await;
        if inner.rates.is_empty() {
            return None;
        }
        let rate = lookup(&inner.rates, model)?;
        Some(cost_from_rate(rate, input_tokens, output_tokens, cache_read_input_tokens))
    }

    /// Refresh the catalogue from cost-core when it is empty or past the TTL. A
    /// fetch failure leaves the cache as-is (and `fetched_at` unset on a cold
    /// cache, so the next call retries) — never poisons it with bad data.
    async fn ensure_fresh(&self) {
        let Some(base) = self.base_url.as_ref() else {
            return;
        };
        {
            let inner = self.inner.read().await;
            if !inner.rates.is_empty() {
                if let Some(at) = inner.fetched_at {
                    if at.elapsed() < REFRESH_TTL {
                        return;
                    }
                }
            }
        }
        let url = format!("{base}/api/v1/pricing");
        match self.http.get(&url).timeout(FETCH_TIMEOUT).send().await {
            Ok(resp) if resp.status().is_success() => match resp.json::<PricingResponse>().await {
                Ok(parsed) => {
                    let mut rates = parsed.rates;
                    for rate in &mut rates {
                        rate.model = rate.model.trim().to_lowercase();
                    }
                    let mut inner = self.inner.write().await;
                    inner.rates = rates;
                    inner.fetched_at = Some(Instant::now());
                }
                Err(error) => tracing::warn!(%error, "pricing: failed to parse catalogue"),
            },
            Ok(resp) => tracing::warn!(status = %resp.status(), "pricing: catalogue fetch non-2xx"),
            Err(error) => {
                tracing::warn!(%error, "pricing: cost-core unreachable; cost stays null until reachable");
            }
        }
    }
}

/// Price a token breakdown against a resolved rate. Pure and synchronous so
/// the cache-tier discount formula is unit-testable without the async
/// HTTP-backed catalogue (`PricingCache::cost_usd` is the async wrapper that
/// resolves `rate` via [`lookup`] and calls this).
///
/// `cache_read_input_tokens` is a leg already folded into `input_tokens` (see
/// `inference-core`'s `total_input_tokens`) — passing 0 reproduces the
/// pre-cache-pricing result exactly, byte for byte. It is billed at
/// `CACHE_READ_DISCOUNT` (0.1x) of the base input rate instead of the full
/// rate: `input` is charged in full first, then the cache-read leg's
/// over-charge is credited back.
fn cost_from_rate(rate: &Rate, input_tokens: i64, output_tokens: i64, cache_read_input_tokens: i64) -> f64 {
    // Token counts comfortably fit in u32; the u32→f64 conversion is
    // lossless (avoids clippy::cast_precision_loss on a raw i64 cast).
    let input = f64::from(u32::try_from(input_tokens.max(0)).unwrap_or(u32::MAX));
    let output = f64::from(u32::try_from(output_tokens.max(0)).unwrap_or(u32::MAX));
    let cache_read = f64::from(u32::try_from(cache_read_input_tokens.max(0)).unwrap_or(u32::MAX));
    input / 1_000_000.0 * rate.input_per_million
        + output / 1_000_000.0 * rate.output_per_million
        - cache_read / 1_000_000.0 * rate.input_per_million * (1.0 - CACHE_READ_DISCOUNT)
}

/// Resolve a model name to a rate: exact key, else the longest prefix key, else
/// the `default` row.
fn lookup<'a>(rates: &'a [Rate], model: &str) -> Option<&'a Rate> {
    let key = model.trim().to_lowercase();
    if !key.is_empty() {
        if let Some(rate) = rates.iter().find(|r| r.model == key) {
            return Some(rate);
        }
        let mut best: Option<&Rate> = None;
        let mut best_len = 0usize;
        for rate in rates {
            if rate.model == DEFAULT_MODEL_KEY {
                continue;
            }
            if key.starts_with(&rate.model) && rate.model.len() > best_len {
                best = Some(rate);
                best_len = rate.model.len();
            }
        }
        if best.is_some() {
            return best;
        }
    }
    rates.iter().find(|r| r.model == DEFAULT_MODEL_KEY)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn rates() -> Vec<Rate> {
        vec![
            Rate {
                model: "default".into(),
                input_per_million: 3.0,
                output_per_million: 15.0,
            },
            Rate {
                model: "gpt-4o-mini".into(),
                input_per_million: 0.15,
                output_per_million: 0.60,
            },
            Rate {
                model: "claude-sonnet".into(),
                input_per_million: 3.0,
                output_per_million: 15.0,
            },
        ]
    }

    #[test]
    fn cache_read_tokens_zero_reproduces_the_pre_cache_pricing_result() {
        // Purely additive: a caller that passes 0 cache tokens (every caller
        // before this change, and every non-Anthropic provider today) must
        // get byte-for-byte the same cost as before the cache-tier discount
        // existed.
        let rate = Rate {
            model: "claude-sonnet".into(),
            input_per_million: 3.0,
            output_per_million: 15.0,
        };
        let plain = 1_000.0 / 1_000_000.0 * 3.0 + 500.0 / 1_000_000.0 * 15.0;
        assert!((cost_from_rate(&rate, 1_000, 500, 0) - plain).abs() < 1e-12);
    }

    #[test]
    fn cache_read_tokens_are_billed_at_a_tenth_of_the_input_rate() {
        // 8_520 input_tokens folds in an 8_000-token cache read (Anthropic's
        // `total_input_tokens` fold-in). Only 520 tokens are fresh input;
        // the other 8_000 are billed at CACHE_READ_DISCOUNT (0.1x) of the
        // $3/1M input rate, not the full rate.
        let rate = Rate {
            model: "claude-sonnet".into(),
            input_per_million: 3.0,
            output_per_million: 15.0,
        };
        let got = cost_from_rate(&rate, 8_520, 42, 8_000);
        let want = 520.0 / 1_000_000.0 * 3.0 // fresh input, full rate
            + 8_000.0 / 1_000_000.0 * 3.0 * 0.1 // cached input, 10x discount
            + 42.0 / 1_000_000.0 * 15.0; // output, unaffected
        assert!(
            (got - want).abs() < 1e-12,
            "got {got}, want {want} (a naive full-rate charge on all 8,520 input tokens would be {})",
            8_520.0 / 1_000_000.0 * 3.0 + 42.0 / 1_000_000.0 * 15.0
        );
    }

    #[test]
    fn exact_and_prefix_and_default() {
        let r = rates();
        assert_eq!(lookup(&r, "gpt-4o-mini").unwrap().model, "gpt-4o-mini");
        assert_eq!(
            lookup(&r, "claude-sonnet-4-6").unwrap().model,
            "claude-sonnet"
        );
        assert_eq!(lookup(&r, "totally-unknown").unwrap().model, "default");
    }

    #[tokio::test]
    async fn disabled_cache_returns_none() {
        let cache = PricingCache::new(None, reqwest::Client::new());
        assert!(cache
            .cost_usd("gpt-4o-mini", 1000, 500, 0, 0)
            .await
            .is_none());
    }
}
